internal/app/app.go

```diff
@@ -0,0 +1,153 @@
+	ctx    context.Context
+	cancel context.CancelCauseFunc
+
+	mu sync.Mutex
+}
+
+type options struct {
+	timeout time.Duration
+	initCtx context.Context
+}
+type option func(*options)
+
+// App is the main application object.
+type App struct {
+	conn    *dbus.Conn
+	obj     *busObject
+	timeout time.Duration
+}
+
+// Apply is a function called via D-Bus to apply the system proxy settings.
+func (b *busObject) Apply(http, https, ftp, socks, no, mode string) *dbus.Error {
+	// Methods calls spin up separate goroutines, so ensure we don't run them in parallel
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	log.Infof("Called Apply(%s, %s, %s, %s, %s, %s)", http, https, ftp, socks, no, mode)
+
+	p, err := proxy.New(b.ctx, http, https, ftp, socks, no, mode)
+	// We need to cancel the context in a deferred function to get the final
+	// state of the error variable.
+	defer func() { b.cancel(err) }()
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	if b.ctx.Value(DryRun{}) == true {
+		return nil
+	}
+
+	err = p.Apply()
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	return nil
+}
+
+// New creates a new App object.
+func New(opts ...option) (a *App, err error) {
+	defer decorate.OnError(&err, "cannot initialize application")
+
+	// defaults
+	args := options{
+		timeout: 1 * time.Second,
+		initCtx: context.Background(),
+	}
+	// applied options
+	for _, o := range opts {
+		o(&args)
+	}
+
+	// Don’t call dbus.SystemBus which caches globally system dbus (issues in tests)
+	conn, err := dbus.ConnectSystemBus()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancelCause(args.initCtx)
+	obj := busObject{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	if err = conn.Export(&obj, dbusObjectPath, dbusInterface); err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+	if err = conn.Export(introspect.NewIntrospectable(&introspect.Node{
+		Name: dbusObjectPath,
+		Interfaces: []introspect.Interface{
+			introspect.IntrospectData,
+			{
+				Name:    dbusInterface,
+				Methods: introspect.Methods(&obj),
+			},
+		},
+	}), dbusObjectPath, introspect.IntrospectData.Name); err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+
+	reply, err := conn.RequestName(dbusInterface, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		_ = conn.Close()
+		return nil, fmt.Errorf("D-Bus name already taken")
+	}
+
+	return &App{
+		conn:    conn,
+		obj:     &obj,
+		timeout: args.timeout,
+	}, nil
+}
+
+// Wait blocks until the context is cancelled or the timeout is reached,
+// returning an appropriate exit code and closing the D-Bus connection before
+// exiting.
+func (a *App) Wait() int {
+	defer a.conn.Close()
+
+	for {
+		select {
+		case <-time.After(a.timeout):
+			log.Infof("No signal received, exiting")
+			return 0
+		case <-a.obj.ctx.Done():
+			if err := context.Cause(a.obj.ctx); err != nil && !errors.Is(err, context.Canceled) {
```

Comment:
To test: if you pass a parent context and cancel this one (so timeout or returned value), what do you get in context.Cause? "parent context cancelled" or anything else?

Also, it seems you return success if the context is canceled, which seems counter-intuitive? (I think this is related to you defer above)

Reply:
yep, with Edu we played with multi error join yesterday. I see you are playing with other parts of the new release!

Reply:
but did you try cancelling the parent context? (as I think that will cancel the child context to with the same "Cause")?

Reply:
^ that’s the part I’m wondering about if the parent context is cancelled, this is probably not a "No error" in the main goroutine.

Reply:
For instance, you are ctrl+C in main(), which is triggered when applying was doing something. The associated context (main/parent context) will be cancelled, and this should be an error